pkg/clustermesh/clustermesh.go

```diff
@@ -0,0 +1,154 @@
+// Copyright 2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package clustermesh
+
+import (
+	"fmt"
+
+	"github.com/cilium/cilium/pkg/controller"
+	"github.com/cilium/cilium/pkg/kvstore/store"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/option"
+)
+
+const (
+	// configNotificationsChannelSize is the size of the channel used to
+	// notify a clustermesh of configuration changes
+	configNotificationsChannelSize = 512
+)
+
+// Configuration is the configuration that must be provided to
+// NewClusterMesh()
+type Configuration struct {
+	// Name is the name of the remote cluster cache. This is for logging
+	// purposes only
+	Name string
+
+	// ConfigDirectory is the path to the directory that will be watched for etcd
+	// configuration files to appear
+	ConfigDirectory string
+
+	// NodeKeyCreator is the function used to create node instances as
+	// nodes are being discovered in remote clusters
+	NodeKeyCreator store.KeyCreator
+}
+
+// ClusterMesh is a cache of multiple remote clusters
+type ClusterMesh struct {
+	// conf is the configuration, it is immutable after NewClusterMesh()
+	conf Configuration
+
+	mutex         lock.RWMutex
+	clusters      map[string]*remoteCluster
+	controllers   *controller.Manager
+	configWatcher *configDirectoryWatcher
+}
+
+// NewClusterMesh creates a new remote cluster cache based on the
+// provided configuration
+func NewClusterMesh(c Configuration) (*ClusterMesh, error) {
+	cm := &ClusterMesh{
+		conf:        c,
+		clusters:    map[string]*remoteCluster{},
+		controllers: controller.NewManager(),
+	}
+
+	w, err := createConfigDirectoryWatcher(c.ConfigDirectory, cm)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create config directory watcher: %s", err)
+	}
+
+	cm.configWatcher = w
+
+	cm.controllers.UpdateController("clustermesh-config-fsnotify",
+		controller.ControllerParams{
+			DoFunc: func() error { return cm.configWatcher.watch() },
+		},
+	)
+
+	return cm, nil
+}
+
+// Close stops watching for remote cluster configuration files to appear and
+// will close all connections to remote clusters
+func (cm *ClusterMesh) Close() {
+	if cm.configWatcher != nil {
+		cm.configWatcher.close()
+	}
+	cm.controllers.RemoveAll()
+}
```

Comment:
This is not removing the clusters, so it doesn't call `remoteCluster.onRemove()` on any cluster that is left when this `Close` method is called, so it is leaking controllers (and therefore goroutines, channels, etc.).